### PR TITLE
ARTEMIS-2321 Paging GC improvements

### PR DESCRIPTION
**Pull request by franz1981**
It contains several improvements:

- PageCursorProviderImpl can use primitive maps
- PageCache doesn't need a Page reference
- Removed unnecessary volatile/Atomic operations and fields
- PageCursorInfo caches number of msgs to save Page::read

### CODE REVIEW

**Comment by franz1981** [comment]
The most relevant one from a functional point of view is the non-blocking Page::read on page cache.
I've maintained a separate commit for each improvement to make it easier to review them separately.
Although not really a bug I've recently looked into many stacktraces showing many threads blocked on soft cache due to slow `Page::read`, causing the broker to be unresponsive and suicide itself right after:
this change (together with https://github.com/apache/activemq-artemis/pull/2633) should make the broker less prone to became unresponsive under a storm of page cache misses.

**Comment by franz1981** [comment]
@wy96f Please take a look :+1:

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java [hunk 1]
```diff
@@ -132,43 +137,57 @@ public PagedReference newReference(final PagePosition pos,
...
+               inProgressReadPage = readPage;
+               inProgressReadPages.put(pageId, readPage);
             }
          }
-
-         return cache;
+         if (createPage) {
+            return readPage(pageId, page, cache, inProgressReadPage);
+         } else {
+            return inProgressReadPage.get(10, TimeUnit.SECONDS);
```
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: this fixed timeout here is bad. I really like the logic...

However:

i - it should be configurable somehow. (it would be nice if there is already another value that would semantic apply in our configuration, but I don't think that will be the case here).
ii - a Timeout should be considered an IOException, and CriticalIOListener should be called on this case.

**Comment by clebertsuconic** [comment]
Shouldn't these commits be squashed?


My only issue here is with an IOException and a Timeout on reading.

**Comment by franz1981** [comment]
@clebertsuconic I've added detection for slow Page::read, but I'm not sure I'm using correcty the `CriticalAnalyzer`: let me know wdyt mate :+1:

**Comment by franz1981** [comment]
I'm sure I need to add a test to cover slow page reading for both an awaiting thread and a performing one

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java [hunk 2]
```diff
@@ -21,26 +21,26 @@
...
 
-   private static final Logger logger = Logger.getLogger(SoftValueHashMap.class);
+   private static final Logger logger = Logger.getLogger(SoftValueLongObjectHashMap.class);
 
    // The soft references that are already good.
    // too bad there's no way to override the queue method on ReferenceQueue, so I wouldn't need this
    private final ReferenceQueue<V> refQueue = new ReferenceQueue<>();
 
-   private final Map<K, AggregatedSoftReference> mapDelegate = new HashMap<>();
+   private final LongObjectMap<AggregatedSoftReference<V>> mapDelegate = new LongObjectHashMap<>();
```
Comment (franz1981) [inline artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: @michaelandrepearce Do you think it could be a problem to use it directly? I remember you've recently working with some primitive map of Netty...

**Comment by franz1981** [comment]
@qihongxu @wf96q it should improve paging scalability over the same address (with topic for sure) a lot, especially with small heaps (and GC is cleaning up many requested PageCaches)

**Comment by franz1981** [comment]
@clebertsuconic I have added a test using Mockito, but I was planning to add some more to cover some weird concurrent corner case too

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java (hunk 2 above)
Comment (michaelandrepearce) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: Why do you need to use it directly? They implement map interface. And the add primitive methods that will be invoked at runtime, you dont need to make the field the concrete impl. This is the trick all primitive impl use.

**Comment by michaelandrepearce** [comment]
Added comment inline

File: artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java (hunk 2 above)
Comment (franz1981) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: But I wanted to avoid the boxing/unboxing cost: the page cache (that uses this soft value map) is sensible to GC cycles and the more is your allocation rate, the more likely you will have a minor GC, making the cache to clean some of its entries...
Avoiding boxing/unboxing is just a way to let the entries stays much longer and makes this map more effective.
TBH I don't know (the details I mean) why we are using a soft value map here: I would have used a LRU map instead with a fixed size of entries instead like https://github.com/real-logic/agrona/blob/master/agrona/src/main/java/org/agrona/collections/IntLruCache.java : need to ask @clebertsuconic about it...

Comment (michaelandrepearce) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: It dossnt box and unbox. Thats the trick they use. By implementing the same interface method but with primitive param at runtime the primitive version is invoked.

Comment (franz1981) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: Probably I've misunderstood your comment Michael, sorry.
I have understood that if you:
```
Map<Long, V> map;
...
V put(Long k, V value) {
    return map.put(k, value);
}

```
In this case a caller that will invoke put(...) is forced to provide a Long, instead of a primitive `long`...no? Or am I missing something?

Comment (michaelandrepearce) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: SoftValue is so it can gc. But if it doesnt has to can keep as much in memory as possible. Rather than a strict bound

Comment (michaelandrepearce) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: Youre missing something. You missing the wonder of the jvm at runtime. It like you have a method in interface that takes object. Your impl has to support that but it can implement the same name with a tighter typing and that will be invoked instead.

Comment (franz1981) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: That's clear to me, but let me explain better what I mean...
If we have:
```
private final LongObjectMap<AggregatedSoftReference<V>> mapDelegate
```
and externally we provide:
```
//(1)
V put(long k, V value) {
    return mapDelegate.put(k, value);
}

//(2)
V put(Long k, V value) {
    return mapDelegate.put(k, value);
}
```
If caller will call (1), the javac will hide a boxing (long -> Long) to allow calling `map.put(k, value)` given that 
mapDelegate has been declared as `Map<Long,...`.

Comment (michaelandrepearce) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: No it doesnt

Comment (franz1981) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: Try to debug it and you will tell me bud :)
Or look at the bytecode ;)

Comment (michaelandrepearce) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: So i think im missing something here, if i have two methods, the tighter typed one will get invoked.

Comment (franz1981) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: @michaelandrepearce 
Try to put a debug print on `Long::valueOf(long)` using:
```
   private static final Map<Long, Object> primitiveMap = new LongObjectHashMap<>();

   public static void main(String[] args) {
      final Object value = new Object();
      for (long l = 0 ; l<10000;l++) {
         final long x = l^l;
         primitiveMap.put(x, value);
      }
   }
```

Comment (michaelandrepearce) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: Ive realised what ive done, ive missread 

"LongObjectMap" as "LongObjectHashMap" ... you using LongObjectMap as field value not LongObjectHashMap.

Comment (michaelandrepearce) [reply artemis-core-client/src/main/java/org/apache/activemq/artemis/utils/SoftValueLongObjectHashMap.java]: youre all good.

**Comment by michaelandrepearce** [comment]
LGTM

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java [hunk 3]
```diff
@@ -132,43 +156,88 @@ public PagedReference newReference(final PagePosition pos,
...
             }
          }
+         //We need to create a new criticalComponent each time, because
+         //each read page action can happen from a different thread
+         final CriticalComponent readPageActionComponent;
+         if (criticalAnalyzer == null) {
+            readPageActionComponent = null;
+         } else {
+            final SimpleString address = pagingStore.getAddress();
+            readPageActionComponent = new CriticalComponentImpl(criticalAnalyzer, PAGE_CURSOR_PROVIDER_CRITICAL_PATHS) {
```
Comment (wy96f) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: Do we need to cache the CriticalComponentImpl here to reduce memory footprint?

Comment (franz1981) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: I have no idea, maybe yes, but I don't want to make this code more complex then necessary: consider that on each `Page::read` request we already create 1 `CompleableFuture` and each completed read will produce several Paged messages as well: saving just 1 instance of CriticalComponentImpl maybe is not a super-win...
I'm more concerned from a functional point of view instead:  in this version I'm triggering the critical analyzer in both cases ie on Page::read and on who is waiting for it. 
Probably I can just leave the critical analyzer to monitor Page:.read and just logging otherwise.. @clebertsuconic wdyt?

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: I would expect the actual component to implement CriticalComponent

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: @franz1981 the implementation here is wrong I think..

you should have PageCursorProviderImpl extending CriticalComponentImpl.

Add it upon construction.. (check remove points when the address is removed)

and just call enter / leave...


like all the other implementations.

Comment (franz1981) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: I wasn't sure about it because you can have several concurrent enterCritical() calls on the same component,while I want the critical analyser to report which pageId read was too slow... 
Wdyt?

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: It wasn't engineered for that.. it would be too much information.

if the Critical Analyzer kicks in (remember, it wasn't supposed to), the relevant information is the stack trace and the methods where you have a possible deadlock.

Comment (franz1981) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: I was thinking to let Page extends CriticalComponentImpl and wrap the read with enter/exit..leaving the CompleteableFuture without it (it is not needed we monitor page read), given that we don't need to monitor both (will leave just the timeout logs).
I know that is not a perfect fit for the critical analyser ..but it could work?

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: @franz1981 you are reinventing things on the CriticalAnalyzer. I wouldn't go that route. Just stick to what's already there.

The CriticalAnalyzer is already complicated enough :)

The CriticalAnalyzer was intended to capture deadlocks. upon a critical call you call enter / leave on a finalize call. I strongly recommend against anything different.

it could work? I don't know.. you're using it in a different way than originally written, and I would not have enough time to test it.

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: The original design expected the component on the list as long as it lives. you are for instance adding and removing it like crazy on each write. I don't have an idea of what would be the implications. Definitely not being used as intended.

Comment (franz1981) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: Got it , but given they concurrent behaviour of such component if you have 1 slow page read and all the others fast enough, the slowest one won't be detected..
I have already a test in this PR to detect slow page read without concurrenct requests: I will create a new one just to check if my guess is correct, thanks!

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: @franz1981 yes, you're right... it won't work with multiple Threads.. I don't know what to do on that case.

perhaps we need multiple Components.. but I would on that case use a ThreadLocal or some other cache. I don't know.

Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: Perhaps implement something on the PageCursor where isExpired would return false, and have individual measures using some sort of thread local?

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java [hunk 4]
```diff
@@ -132,52 +146,99 @@ public PagedReference newReference(final PagePosition pos,
...
+                            pageId, pagingStore.getAddress(), elapsedMillis);
+            }
+         }
          page.open();
-
+         final long startedReadPage = System.nanoTime();
          List<PagedMessage> pgdMessages = page.read(storageManager);
+         final long elapsedReadPage = System.nanoTime() - startedReadPage;
+         if (elapsedReadPage > PAGE_READ_TIMEOUT_NS) {
+            logger.warnf("Page::read for pageNr=%d on cursor %s tooks %d ms to read %d bytes", pageId,
```
Comment (franz1981) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: @clebertsuconic @jbertram 
In order to not modify `Page::read` and introduce a whole new "failure" state (half pages read with critical IO error raised) 
I've preferred to add some more detailed logs in order to detect weird behaviours, long wait or `page-max-concurrent-io` misconfigurations.
I suppose that adding a proper "slowness disk/JVM suicide" deserve its own PR.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java (hunk 1 above)
Comment (clebertsuconic) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: I thought you were going to fix this? Is there any extra commits you did not push?

Comment (franz1981) [reply artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java]: It is already outdated: I've added proper logging around this method and the behaviour is the same as before

**Comment by franz1981** [comment]
I've moved 2 commits re scalability of paging cache from here into https://github.com/apache/activemq-artemis/pull/2667
